modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionContext.java

```diff
@@ -242,6 +247,30 @@ public void execute(RunnableX task, Consumer<Throwable> onError) {
         });
     }
 
+    /**
+     * Submits a Runnable task for execution and returns a Future
+     * representing that task. The Future's {@code get} method will
+     * return {@code null} upon <em>successful</em> completion.
+     *
+     * @param task the task to submit.
+     * @return a Future representing pending task
+     */
+    public Future<?> submit(RunnableX task, Consumer<Throwable> onError) {
+        if (isCancelled())
```

Comment:
no task should be scheduled after the context is canceled, so lets verify this with an assertion